sys/kern/imgact_elf.c

```diff
@@ -789,139 +809,16 @@ __elfN(load_section)(const struct image_params *imgp, vm_ooffset_t offset,
 	return (0);
 }
 
-#ifdef __ELF_CHERI
-/*
- * Note: this needs to be called before load_sections() since we need to know
- * the correct base before mapping the segments.
- */
-static int
-__elfN(ensure_representable_base)(const char *execpath, const Elf_Ehdr *hdr,
-    const Elf_Phdr *phdr, Elf_Addr *preferred_rbase)
-{
-	/* Initialize start_addr so that MIN() produces something useful. */
-	Elf_Addr start_addr = ~0UL;
-	Elf_Addr max_addr = 0;
-	size_t total_size;
-	Elf_Addr adjusted_base;
-
-	for (size_t i = 0; i < hdr->e_phnum; i++) {
-		u_long seg_size, seg_addr, end_addr;
-
-		if (phdr[i].p_type != PT_LOAD || phdr[i].p_memsz == 0)
-			continue;
-
-		seg_addr = trunc_page(phdr[i].p_vaddr);
-		seg_size = round_page(phdr[i].p_memsz + phdr[i].p_vaddr -
-		    seg_addr);
-		end_addr = seg_addr + seg_size;
-
-		start_addr = MIN(start_addr, seg_addr);
-		max_addr = MAX(max_addr, end_addr);
-	}
-
-	/*
-	 * Ensure that we can represent a capability spanning the entire
-	 * mapping and move it if not (only possible for PIE).
-	 */
-	total_size = max_addr - start_addr;
-	if (__predict_true(*preferred_rbase ==
-	    CHERI_REPRESENTABLE_BASE(*preferred_rbase, total_size))) {
-		/* No adjustment required */
-		return (0);
-	}
-	if (hdr->e_type == ET_EXEC) {
-		/*
-		 * We can't change the load address for position dependent
-		 *  executables, so we have to give up and report an error.
-		 */
-		uprintf("Warning: Attempted to load position-dependent "
-		    "executable with non-representable base: %s\n", execpath);
-		return (ERANGE); /* XXX: EPRECISION or similar? */
-	}
-	/*
-	 * Note: CHERI_REPRESENTABLE_BASE aligns down, but we have to align
-	 * upwards here to avoid rounding down to zero for the main executable.
-	 * For RTLD we also align upwards to avoid aligning down into the
-	 * memory region for the main binary.
-	 */
-	adjusted_base = roundup2(*preferred_rbase,
```

Comment:
We still need the logic to potentially change rbase(since elfN(load_sections) needs the new value. Probably best to food that into elfN(build_imgact_capability).

Reply:
I see that vm_map_reservation_create_locked does the CHERI_REPRESENTABLE_BASE(), but if that rounds down to zero we can't create the mapping, right?

Reply:
Yeah, I think a sensible way to go is to roundup before calling into `vm_map_reservation_create()`, I'm not sure if we want to add flags or somehow conditionally change the behaviour of reservation_create from rounding down to rounding up. The rounding up does not seem common.